source/unstoppable_baton.cpp

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2019-present Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <unifex/unstoppable_baton.hpp>
+
+namespace unifex::_ub {
+
+void unstoppable_baton::post() noexcept {
+  void* const signalledState = this;
+
+  // replace the stack of waiting operations with a sentinel indicating we've
+  // been signalled
+  void* top = state_.exchange(signalledState, std::memory_order_acq_rel);
+
+  if (top == signalledState) {
+    // we were already signalled so there are no waiting operations
+    return;
+  }
+
+  // We are the first thread to set the state to signalled; iteratively pop
+  // the stack and complete each operation.
+  auto op = static_cast<_op_base*>(top);
+  while (op != nullptr) {
+    std::exchange(op, op->next_)->set_value();
+  }
+}
+
+void unstoppable_baton::start_or_wait(_op_base& op, unstoppable_baton& baton) noexcept {
+  // Try to push op onto the stack of waiting ops.
+  void* const signalledState = &baton;
+
+  void* top = baton.state_.load(std::memory_order_acquire);
+
+  do {
+    if (top == signalledState) {
+      // Already in the signalled state; don't push it.
+      op.set_value();
+      return;
+    }
+
+    // note: on the first iteration, this line transitions op.next_ from
+    //       indeterminate to a well-defined value
+    op.next_ = static_cast<_op_base*>(top);
+  } while (!baton.state_.compare_exchange_weak(
+      top,
+      static_cast<void*>(&op),
+      std::memory_order_release,
+      std::memory_order_acquire));
+}
```

Comment:
This is a static function that takes the baton *second* because the caller will have `op` in the first argument position before calling.  If this were a member function, I'd expect the compiler to have to emit instructions to juggle registers before calling to match the calling convention.  I would assume that register renaming would make the juggling zero runtime cost, but, assuming the instructions are emitted, it's not zero binary size cost.

Reply:
Have you veriifed that the compiler does not emit the register juggling instructions when calling this?
If so, maybe record the intent/rationale in a comment (either here or at the declaration)?

Reply:
I hadn't done it, but now I have: https://godbolt.org/z/nej31v.

At `-O3`, the differences seem negligible because of the effects of inlining, but at `-Oz`, `unifex::_amre::_op_base::start()` is actually emitted, and it's one instruction shorter in the case of the static method.

Interestingly, the implementation of `start_or_wait` is also slightly shorter at `-Oz` when it's a static function (change the `#if 0` to `#if 1` at the Godbolt to see the differences).  Fooling around a little more, it seems I should put `next_` and `setValue_` first and second (in either order) to shrink `start_or_wait` because ARM can `ldp` the first two fields into a pair of registers.